### PR TITLE
[WIP] Begin frontend functionality for Add/Edit Case

### PR DESCRIPTION
**Pull request by wittejm**
So far includes responsive "add case" and  "edit case" buttons to show the edit panel, and panel includes responsive buttons and form validation.

Redux/axios not touched yet.

### CODE REVIEW

**Comment by wittejm** [comment]
The current commit has a responsive panel and should have (nearly?) all the behavior prior to processing an actual edit.

A few differences from the mockup are:
- The edit-case and add-charge buttons are hidden while the case is being edited.
- The Case Number now only appears in the Case header, since we're not making it editable.
- The buttons on the panel are different for Add vs Edit case: Create/Cancel vs Update/Remove/Cancel
- Form validation is "live" -- it's impossible to type values into Balance that aren't of "###.##" format or Birth Year that aren't "####' format. We can eliminate a few validation error banners this way.

@hmarcks thoughts on these differences?

add_case
editing_panel

**Comment by hmarcks** [comment]
Looking good!

> The edit-case and add-charge buttons are hidden while the case is being edited.
Can you explain this? I think we always want to show the buttons.

> The Case Number now only appears in the Case header, since we're not making it editable.
Sounds good

> The buttons on the panel are different for Add vs Edit case: Create/Cancel vs Update/Remove/Cancel
Oh? That's the same in the mockups

> Form validation is "live" -- it's impossible to type values into Balance that aren't of "###.##" format or Birth Year that aren't "####' format. We can eliminate a few validation error banners this way.
Sounds good, only concern with no validation message for Birth Year is that it may be confusing if you don't enter the correct year format (2 numbers) then aren't sure why it's not submitting.

**Comment by wittejm** [comment]
> Can you explain this? I think we always want to show the buttons.

The example header pre-click and post-click:
pre-edit
clicked-edit

If the edit button stays there, what should it do? I think it should either become inactive or disappear. The expected behavior on click isn't obvious as to whether it causes an update or cancel.

If the +charge button stays, that lets the user open the add/edit-case panel and the add-charge panel at the same time. I could go either way on this but lean toward allowing only one at a time, so that the user must create an empty case and them populate it with charges. If the user can edit a case, and then add charges to it simultaneously, they might get confused about the expected workflow: in that "updating a case" might mean either 1) "update case info and all contained charges also, if there are pending charge edits" or that it would mean 2) "just update case info, and leave the charge changes pending."
 

> Oh? That's the same in the mockups

Ah yep. I was only looking at the final mockup : https://github.com/codeforpdx/recordexpungPDX/issues/1079#issuecomment-627601420

> Sounds good, only concern with no validation message for Birth Year is that it may be confusing if you don't enter the correct year format (2 numbers) then aren't sure why it's not submitting.

Good point. I can keep that validation message too.

**Comment by hmarcks** [comment]
Oh got it, sorry I was thinking of the submit/remove/cancel buttons. I agree we should limit to one action at a time so hiding the buttons should work well. Thanks

File: src/backend/expungeservice/record_creator.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from functools import lru_cache
 from itertools import product, groupby
 from typing import List, Dict, Tuple, Any, Callable, Type
-from datetime import datetime
+from datetime import date, datetime
```
Comment (wittejm) [inline src/backend/expungeservice/record_creator.py:5]: will undo

**Comment by wittejm** [comment]
Latest commit covers Add and Update case. Still todo are Undo changes and Remove.

File: src/frontend/src/redux/search/reducer.ts [hunk 2]
```diff
@@ -40,14 +43,24 @@ export function searchReducer(
       }
       return {...state, questions: questions, loading: true};
     case ANSWER_DISPOSITION:
-      const edits = JSON.parse(JSON.stringify(state.edits));
+      edits = JSON.parse(JSON.stringify(state.edits));
```
Comment (KentShikama) [inline src/frontend/src/redux/search/reducer.ts:46]: Why remove the const here? `edits` can just be a different variable with the same name across the two cases.

File: src/backend/expungeservice/record_creator.py [hunk 3]
```diff
@@ -149,11 +149,33 @@ def _edit_search_results(search_result_cases: List[OeciCase], edits) -> Tuple[Li
             if case_number in edits.keys():
                 if edits[case_number]["action"] == "edit":
                     edited_case, new_charges = RecordCreator._edit_case(case, edits[case_number])
-                    edited_cases.append(edited_case)
+                    edited_cases.append(replace(edited_case, summary=replace(edited_case.summary, edit_status=EditStatus.EDITED)))
```
Comment (KentShikama) [inline src/backend/expungeservice/record_creator.py:152]: This should be done in edit case:

`edited_summary = replace(case.summary, edit_status=EditStatus.EDITED, **case_summary_edits)`

File: src/frontend/src/redux/search/reducer.ts (hunk 2 above)
Comment (wittejm) [reply src/frontend/src/redux/search/reducer.ts:46]: Ah, seems that case blocks don't have a new scope unless I make a new `{}` block. Adding that to allow two `const`s seems fine.

**Comment by wittejm** [comment]
Pushed changes for Remove case and Undo.

There's a confirm message for hitting Remove on a manually added case. I wasn't sure whether to do the same for hitting U

**Comment by wittejm** [comment]
n

**Comment by wittejm** [comment]
Also todo: don't pre-populate balance and DOB fields when Adding new case.

**Comment by KentShikama** [comment]
Should we close this?

**Pull request closed without merging**